subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -2110,17 +2119,34 @@ public ArtifactsResolved(ResolverResults results) {
     }
 
     private DefaultArtifactCollection artifactCollection(AttributeContainerInternal attributes, Spec<? super ComponentIdentifier> componentFilter, boolean lenient, boolean allowNoMatchingVariants, boolean selectFromAllVariants) {
-        ImmutableAttributes viewAttributes = attributes.asImmutable();
+        AttributeContainerInternal viewAttributes = attributes;
+        // If the user desires legacy behavior, we need to make sure that the view attributes are a separate collection so
+        // they do not pick up later attribute changes to the configuration.
+        if (useLegacyAttributeSnapshottingBehavior()) {
+            viewAttributes = viewAttributes.asImmutable();
+        }
+
         DefaultResolutionHost failureHandler = new DefaultResolutionHost();
         ResolutionBackedFileCollection files = new ResolutionBackedFileCollection(
             new SelectedArtifactsProvider(Specs.satisfyAll(), viewAttributes, componentFilter, allowNoMatchingVariants, selectFromAllVariants, new VisitedArtifactsSetProvider()), lenient, failureHandler, taskDependencyFactory
         );
         return new DefaultArtifactCollection(files, lenient, failureHandler, calculatedValueContainerFactory);
     }
 
-    @Override
-    public ConfigurationRole getRoleAtCreation() {
-        return roleAtCreation;
+    private static boolean useLegacyAttributeSnapshottingBehavior() {
+        if (useLegacyAttributeSnapshottingBehavior == null) {
+            useLegacyAttributeSnapshottingBehavior = Boolean.getBoolean(USE_LEGACY_ATTRIBUTE_SNAPSHOT_BEHAVIOR);
+        }
+
+        if (useLegacyAttributeSnapshottingBehavior) {
+            DeprecationLogger.deprecateSystemProperty(USE_LEGACY_ATTRIBUTE_SNAPSHOT_BEHAVIOR)
+                .withAdvice("Please remove this flag and use the current default behavior.")
+                .willBeRemovedInGradle9()
+                .withUpgradeGuideSection(8, "legacy_attribute_snapshotting")
+                .nagUser();
+        }
+
+        return useLegacyAttributeSnapshottingBehavior;
```

Comment:
We can't do this. We can't have magic system properties for this kind of thing. 

We have a few options:

1. We switch to the new behavior and accept whatever breaks.
2. We switch to the new behavior and mitigate what breaks by working around what plugins might be relying on.
3. We use the new behavior when a feature preview is enabled.
4. We support both behaviors with an API.
5. We only support the existing behavior. (I think we've already ruled this out)

It's hard for me to tell from the PR how big of an impact (1) or (2) would be.

Reply:
Removing this to go with option 1) per our discussion.